{% extends "layout.html" %}
{% load staticfiles %}
{% block title %}FAQ{% endblock %}
{% block description %}FAQ{% endblock %}

{% block main_container %}
<div id="main_container" class="span10">
  <section id="main">

    <div class="container">

      <div class="row-fluid">
        <div class="span12">
          <div class="content-box transparent">
            <div class="padded">
              <h2>Frequently Asked Questions (FAQ)</h2>
            </div>
          </div>
        </div>
      </div>

      <div class="content-box">
        <div class="content-box-header">
          <i class="icon-question-sign"></i> FAQ
        </div>

        <div class="row-fluid">
          <div class="span12">
            <div class="padded">

              <li>
                <h4>How are risk values calculated?</h4>
                <p>See <a href="{% static 'images/snaps/methods.png' %}">Methods</a>.</p>
              </li>

              <hr>
              <li>
                <h4>What is RR (Relative Risk)?</h4>
                <p>See <a href="{% static 'images/snaps/methods.png' %}">Methods</a>.</p>
              </li>

              <hr>
              <li>
                <h4>What is OR (Odds Ratio)?</h4>
                <p>See <a href="{% static 'images/snaps/methods.png' %}">Methods</a>.</p>
              </li>

              <hr>
              <li>
                <h4>What is VCF?</h4>
                <p><a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41" target="_blank">VCF (Variant Call Format)</a> is a typical data format for DNA sequencing data from NGS.</p>
                <p>Example:</p>
<pre class="prettyprint">
##fileformat=VCFv4.1
...
#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                FORMAT      NA00001
20     14370   rs6054257 G      A       29   PASS   NS=3;DP=14;AF=0.5   GT:GQ:DP:HQ 0|0:48:1:51,51
20     17330   .         T      A       3    q10    NS=3;DP=11;AF=0.017 GT:GQ:DP:HQ 0|0:49:3:58,50
</pre>
              </li>

              <hr>
              <li>
                <h4>What is VCF (Whole Genome) or VCF (TruSeq Exome)?</h4>
                <p>VCF file for <b>Whole Genome</b> contains whole (all) the genomic positions of your genome.<p>
                <p>VCF file for <b>TruSeq Exome</b> contains targetted region of <a href="http://www.illumina.com/products/truseq_exome_enrichment_kit.ilmn" target="_blank">TruSeq Exome Enrichment Kit (Illumina, Inc.)</a> of your genome.<p>
              </li>

              <hr>
              <li>
                <h4>What is 23andMe?</h4>
                <p><a href="https://www.23andme.com/" target="_blank">23andMe (23andMe, Inc.)</a> is a common direct-to-consumer genetic testing service .</p>
              </li>

              <hr>
              <li>
                <h4>What does <code>na</code> in genotypes mean?</h4>
                <p><code>na</code> means genotype is Not Available in uploaded file, like following:</p>
<pre class="prettyprint">
if snp is not in uploaded_file:

  if file_format == &#039;23andme&#039;:
      -&gt; na
  elif file_format == &#039;vcf_whole_genome&#039;:
      -&gt; ref
  elif file_format == &#039;vcf_exome_truseq&#039;:
      if is_in_truseq:
          -&gt; ref
      else:
          -&gt; na</pre>
              </li>

              <hr>
              <li>
                <h4>[Advanced] Why have I genotyped-records which is not in my targeted region?</h4>
                <p>Even though your VCF is generated by exome targetted sequencing, some records which is in external exome region may exist.
                  Those records may be artifacts of read-mapping or variant-calling, or matter of exome-targetting in DNA sampling.
                  Nevertheless, they are in VCF because they are passed filter of quality of variant-calling.
                  So, in perGENIE, as far as variant records appear in VCF, they are treated as existing variatns, though they are not in its targeted region.
                </p>
              </li>

              <hr>
              <li>
                <h4>[Advanced] How does the cover rate of GWAS Cataog for each genome file calculated?</h4>
                <p>= 100 * (# of (unique SNPs in GWAS Catalog && available in your genome file)) / (# of unique SNPs in GWAS Catalog)</p>
              </li>

              <hr>
              <li>
                <h4>[Advanced] How does the cover rate of whole genome for each genome file calculated?</h4>
                <p>= 100 * (length of targeted region) / (length of whole genome)</p>
                <p>For VCF of whole genome sequencing, this value is 100%</p>
                <ul>
                  <li>For SNP array of 23andMe, targeted region is considered as the total number of SNPs on array.</li>
                  <li>For TruSeq-Exome-Targeted-Region, targeted region is considered as the whole reigion of targeted region.</li>
                </ul>
              </li>

              <hr>
              <li>
                <h4>[Advanced] Are reference alleles always non-risk allele?</h4>
                <p>No. Reference allele is not always non-risk allele. There are some records that reference allele is risk-allele.
                  Especially, if the frequency of risk-allele is not low, e.g., 0.6 or something, risk-allele maybe the reference allele.
                  Though it is confusable, as it depends on the definition of the risk-allele in each GWAS, it is unavoidable.
                </p>
              </li>

              <hr>
              <li>
                <h4>[Advanced] Why are there no riks values betwee 0 to 1 in charts?</h4>
                <p>まず，Relative risk (RR)は2つの確率の比であるため，常に正の値になります．</p>
                <p>RRが1より大きい値xをとるとき，x倍リスクが高いという意味です．
                  しかし，RRが1より小さい値をとるとき，0.y倍リスクが高い，となりますが，
                  これは逆数をとって，y倍リスクが「低い」と言い換えた方がわかりやすいと考えられます．
                </p>
                <p>そのため，perGENIEではRRが0から1の値を取るときその逆数をとり，
                  また符号をマイナスにしてリスクが低くなるという意味で表記しているので，
                  0から1の値が表示されていません．
                </p>
              </li>


            </div>
          </div>
        </div>
      </div>

    </div>

  </section>
</div>
{% endblock %}

{% block js %}
{% endblock %}
